src/unix/macos.rs

```diff
@@ -0,0 +1,117 @@
+//! Beginning with MacOS 10.3, `utimensat` is supported by the MacOS, so here, we check if the symbol exists
+//! and if not, we fallabck to `utimes`.
+use crate::FileTime;
+use std::ffi::CString;
+use std::fs::{File, Metadata};
+use std::os::unix::prelude::*;
+use std::path::Path;
+use std::{io, ptr};
+
+pub fn set_file_times(p: &Path, atime: FileTime, mtime: FileTime) -> io::Result<()> {
+    set_times(p, Some(atime), Some(mtime), false)
+}
+
+pub fn set_file_mtime(p: &Path, mtime: FileTime) -> io::Result<()> {
+    set_times(p, None, Some(mtime), false)
+}
+
+pub fn set_file_atime(p: &Path, atime: FileTime) -> io::Result<()> {
+    set_times(p, Some(atime), None, false)
+}
+
+fn utimensat_available() -> io::Result<bool> {
+    let lib_name = CString::new("libSystem.dylib".as_bytes())?;
+    let handle = unsafe { libc::dlopen(lib_name.as_ptr() as *const _, 0) };
+    if handle == ptr::null_mut() {
+        return Err(io::Error::last_os_error());
+    }
+    let symbol_name = CString::new("utimensat".as_bytes())?;
+    let address = unsafe { libc::dlsym(handle, symbol_name.as_ptr() as *const _) };
+    if address == ptr::null_mut() {
+        Ok(false)
+    } else {
+        Ok(true)
+    }
+}
+
+pub fn set_file_handle_times(
+    f: &File,
+    atime: Option<FileTime>,
+    mtime: Option<FileTime>,
+) -> io::Result<()> {
+    if utimensat_available()? {
+        let (atime, mtime) = match get_times(atime, mtime, || f.metadata())? {
+            Some(pair) => pair,
+            None => return Ok(()),
+        };
+        let times = [
+            super::to_timespec(&Some(atime)),
+            super::to_timespec(&Some(mtime)),
+        ];
+        let rc = unsafe { libc::futimens(f.as_raw_fd(), times.as_ptr()) };
```

Comment:
This unfortunately doesn't have the desired effect because although this is never called at runtime on older systems it still is linked in the binary and so the entire binary will fail to load. What'll need to happen is that you'll have to make an indirect function call through the pointer returned by `dlsym` instead of using the `libc` crate here.

I'd recommend using something like [`weak.rs` in rust-lang/rust](https://github.com/rust-lang/rust/blob/60960a260f7b5c695fd0717311d72ce62dd4eb43/src/libstd/sys/unix/weak.rs) which actually existed in this repo [until recently](https://github.com/alexcrichton/filetime/commit/2e494fbcd4b60bd6041bb6067573768c19676fa6)

Reply:
Ah OK, I'll have a look at it tomorrow. Thanks for the pointer!

Reply:
Changed according to your suggestion (thanks again BTW!). Have a look and lemme know if that's what you've had in mind. :-)